docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+
+Based on [this comment](https://github.com/rust-lang/api-guidelines/issues/29#issuecomment-342745898) I think the advice for Crate names should be:
+
+* Infer Crate names from package names where feasible:
+  * A package is the directory in which the Cargo.toml exists.
+* Prefer single (lowercase) words for package names where feasible
+* Use __lowercase with underscore__ (a.k.a "snake_case") to split words in package names if needed
+
+I believe this would result in Crate names (which must be Rust identifiers) being identical to the package name at all times.
+
+## Module Layout
+
+Modules form the basic units of imported code in Rust, so identifying where to split functionality and what to call modules is essential. I suspect this is the right level to divide things like collation vs segmentation etc. but it's also far too early to make hard recommendations about this, so I proposed to leave this issue open for the time being (moving code between modules later should be relatively easy).
+
+**Open Question**: Should we have code in the top-level module of a Crate or always use named sub-modules? This probably ties into what Crates we intend to publish and the granularity of the APIs.
+
+## Shared Internal Code
+
+Rust supports having non-public sub-modules within a Crate, and only the top-level "implicit" module is exposed by default (other modules can be exported by the top level module). See also [File hierarchy - Rust By Example](https://doc.rust-lang.org/rust-by-example/mod/split.html).
+
+### Use Private Modules :: required
+
+While local private functions can just be defined in any module, shared internal functions should always be implemented in a non-exported sub-module. This avoids any misleading dependencies between "public" modules and should help ensure clean code dependencies. Code in internal modules should still be tested and documented thoroughly. It is perfectly expected that we would have multiple internal modules for different aspects of ICU4X.
+
+# Structs and Traits
+
+For ICU4X we should be looking to keep code consistent and very unsurprising, especially for layout and behavior of data structures.
+
+## Private vs Public
+
+Rust offers a compelling model for encapsulating implementation details based on a hierarchical view of privacy; see [Visibility and Privacy - The Rust Reference](https://doc.rust-lang.org/reference/visibility-and-privacy.html). This should allow ICU4X to encapsulate all of its implementation details neatly from applications using the library and avoid too many cases of [Hyrum's Law](https://www.hyrumslaw.com/).
+
+### Minimize user-visible structs :: required
+
+While this sounds a bit obvious, I think it's important to stress that with Rust's privacy model, there is never any need to have an internal type be user-visible, and once a type is user-visible outside a Crate, it's published and should not be changed.
```

Comment:
Yes, and for something like ICU4X, where it's used in core code and people don't want to be having to update it often, it seems like we should strongly prefer to avoid breaking version bumps. I can reword the "should not be changed" to "has a potentially high cost to change"